openedx/core/djangoapps/bookmarks/tasks.py

```diff
@@ -0,0 +1,155 @@
+
+    return blocks_info_dict
+
+
+def _paths_from_data(paths_data):
+    """
+    Construct a list of paths from path data.
+    """
+    from .models import PathItem
+    paths = []
+    for path_data in paths_data:
+        paths.append([
+            PathItem(item['usage_key'], item['display_name']) for item in path_data
+            if item['usage_key'].block_type != 'course'
+        ])
+
+    return [path for path in paths if path]
+
+
+def _paths_equal(paths_1, paths_2):
+    """
+    Check if two paths are equivalent.
+    """
+    if len(paths_1) != len(paths_2):
+        return False
+
+    for path_1, path_2 in zip(paths_1, paths_2):
+        if len(path_1) != len(path_2):
+            return False
+
+        for path_item_1, path_item_2 in zip(path_1, path_2):
+            if path_item_1.display_name != path_item_2.display_name:
+                return False
+
+            usage_key_1 = path_item_1.usage_key.replace(
+                course_key=modulestore().fill_in_run(path_item_1.usage_key.course_key)
+            )
+            usage_key_2 = path_item_1.usage_key.replace(
+                course_key=modulestore().fill_in_run(path_item_2.usage_key.course_key)
+            )
+            if usage_key_1 != usage_key_2:
+                return False
+
+    return True
+
+
+def _update_xblocks_cache(course_key):
+    """
+    Calculate the XBlock cache data for a course and update the XBlockCache table.
+    """
+    from .models import XBlockCache
+    blocks_data = _calculate_course_xblocks_data(course_key)
+
+    def update_block_cache_if_needed(block_cache, block_data):
+        """ Compare block_cache object with data and update if there are differences. """
+        paths = _paths_from_data(block_data['paths'])
+        if block_cache.display_name != block_data['display_name'] or not _paths_equal(block_cache.paths, paths):
+            block_cache.display_name = block_data['display_name']
+            block_cache.paths = paths
+            block_cache.save()
+
+    with transaction.commit_on_success():
+        block_caches = XBlockCache.objects.filter(course_key=course_key)
+        for block_cache in block_caches:
+            block_data = blocks_data.pop(unicode(block_cache.usage_key), None)
+            if block_data:
+                update_block_cache_if_needed(block_cache, block_data)
+
+    for block_data in blocks_data.values():
+        with transaction.commit_on_success():
+            paths = _paths_from_data(block_data['paths'])
+            block_cache, created = XBlockCache.objects.get_or_create(usage_key=block_data['usage_key'], defaults={
+                'course_key': course_key,
+                'display_name': block_data['display_name'],
+                'paths': paths,
+            })
+
+            if not created:
+                update_block_cache_if_needed(block_cache, block_data)
+
+
+@task(name=u'openedx.core.djangoapps.bookmarks.tasks.update_xblock_cache')
+def update_xblocks_cache(course_id):
```

Comment:
It would be nice if we can add some logging in this file, as this is a background task and logging would be really helpful.